acmeCA-2.0: Add exception type to CWPKI2016E

While testing accidentally hit an UnknownHostException, but only the message was included on the error message, making it hard to tell what was wrong, unless you review the FFDC.

For example, this message `CWPKI2016E: The ACME service request for an existing account from the ACME certificate authority at the https://acme-staging-02.api.letsencrypt.org/directory URI failed. The error is 'acme-staging-02.api.letsencrypt.org'.`

when the Exception class really tells you what's going on:

`Caused by: java.net.UnknownHostException: acme-staging-02.api.letsencrypt.org`

For #9017

--- a/dev/com.ibm.ws.security.acme/src/com/ibm/ws/security/acme/internal/AcmeClient.java
+++ b/dev/com.ibm.ws.security.acme/src/com/ibm/ws/security/acme/internal/AcmeClient.java
@@ -1072,7 +1072,7 @@ public class AcmeClient {
 				cause = e;
 			}
 
-			throw new AcmeCaException(Tr.formatMessage(tc, "CWPKI2028E", acmeConfig.getDirectoryURI()), cause);
+			throw new AcmeCaException(Tr.formatMessage(tc, "CWPKI2028E", acmeConfig.getDirectoryURI(), cause), cause);
 		}
 	}
 
@@ -1088,16 +1088,31 @@ public class AcmeClient {
 	@Trivial
 	private static String getRootCauseMessage(Throwable t) {
 		Throwable cause;
-		String rootMessage = t.getMessage();
+		String rootMessage = addExceptionClass(t);
 
 		for (cause = t; cause != null; cause = cause.getCause()) {
 			String msg = cause.getMessage();
 			if (msg != null && !msg.trim().isEmpty()) {
-				rootMessage = msg;
+				rootMessage = addExceptionClass(cause);
 			}
 		}
 
 		return rootMessage;
+	}
+
+	/**
+	 * Add on the Exception class name as sometimes the message does not make sense
+	 * without the name of the Exception included. For example,
+	 * java.net.UnknownHostException: <exampleUnknownHost . net>
+	 * 
+	 * @param t
+	 * @return
+	 */
+	private static String addExceptionClass(Throwable t) {
+		if (t != null) {
+			return t.getClass().getName() + ": " + t.getMessage();
+		}
+		return "";
 	}
 
 	/**

--- a/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/fat/AcmeRevocationTest.java
+++ b/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/fat/AcmeRevocationTest.java
@@ -589,4 +589,60 @@ public class AcmeRevocationTest {
 	private void stopServer(String...msgs) throws Exception {
 		AcmeFatUtils.stopServer(server, msgs);
 	}
+	
+	/**
+	 * Start a pebble server and swap between pebble and boulder and make sure
+	 * we still get a certificate when switching.
+	 * 
+	 * @throws Exception
+	 *             if the test failed for some unforeseen reason.
+	 */
+	@Test
+	public void swapBetweenCAProviders() throws Exception {
+		final String methodName = testName.getMethodName();
+
+		CAContainer pebble = new PebbleContainer();
+		try {
+
+			/*
+			 * Configure and start with Pebble
+			 */
+			ServerConfiguration configuration = originalServerConfig.clone();
+			configureAcmeRevocation(configuration, true);
+			AcmeFatUtils.configureAcmeCA(server, pebble, configuration, false, DOMAIN);
+
+			Log.info(AcmeSimpleTest.class, methodName, "Starting server with Pebble");
+			server.startServer();
+			AcmeFatUtils.waitForSslToCreateKeystore(server);
+			AcmeFatUtils.waitForSslEndpoint(server);
+			AcmeFatUtils.waitForAcmeToCreateCertificate(server);
+			server.setMarkToEndOfLog(server.getDefaultLogFile());
+
+			Certificate[] certificates1 = AcmeFatUtils.assertAndGetServerCertificate(server, pebble);
+
+			Log.info(AcmeSimpleTest.class, methodName, "Swap to Boulder");
+			AcmeFatUtils.configureAcmeCA(server, boulder, configuration, false, DOMAIN);
+			AcmeFatUtils.waitForAcmeToCreateCertificate(server);
+			certificates1 = AcmeFatUtils.waitForNewCert(server, boulder, certificates1);
+			server.setMarkToEndOfLog(server.getDefaultLogFile());
+
+			Log.info(AcmeSimpleTest.class, methodName, "Swap back to Pebble");
+			AcmeFatUtils.configureAcmeCA(server, pebble, configuration, false, DOMAIN);
+			AcmeFatUtils.waitForAcmeToCreateCertificate(server);
+			certificates1 = AcmeFatUtils.waitForNewCert(server, pebble, certificates1);
+
+			Log.info(AcmeSimpleTest.class, methodName, "Swap back to Boulder");
+			AcmeFatUtils.configureAcmeCA(server, boulder, configuration, false, DOMAIN);
+			AcmeFatUtils.waitForAcmeToCreateCertificate(server);
+			certificates1 = AcmeFatUtils.waitForNewCert(server, boulder, certificates1);
+			server.setMarkToEndOfLog(server.getDefaultLogFile());
+
+		} finally {
+			stopServer();
+
+			if (pebble != null) {
+				pebble.stop();
+			}
+		}
+	}
 }

--- a/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/utils/AcmeFatUtils.java
+++ b/dev/com.ibm.ws.security.acme_fat/fat/src/com/ibm/ws/security/acme/utils/AcmeFatUtils.java
@@ -779,8 +779,8 @@ public class AcmeFatUtils {
 	 * @param startingCertificateChain
 	 * @throws Exception
 	 */
-	public static final void waitForNewCert(LibertyServer server, CAContainer acmeContainer, Certificate[] startingCertificateChain) throws Exception {
-		waitForNewCert(server, acmeContainer, startingCertificateChain, SCHEDULE_TIME);
+	public static final Certificate[]  waitForNewCert(LibertyServer server, CAContainer acmeContainer, Certificate[] startingCertificateChain) throws Exception {
+		return waitForNewCert(server, acmeContainer, startingCertificateChain, SCHEDULE_TIME);
 	}
 
 	/**
@@ -791,7 +791,7 @@ public class AcmeFatUtils {
 	 * @param timeout
 	 * @throws Exception
 	 */
-	public static final void waitForNewCert(LibertyServer server, CAContainer acmeContainer, Certificate[] startingCertificateChain, long timeout) throws Exception {
+	public static final Certificate[]  waitForNewCert(LibertyServer server, CAContainer acmeContainer, Certificate[] startingCertificateChain, long timeout) throws Exception {
 		Certificate[] endingCertificateChain;
 		long startTime = System.currentTimeMillis();
 		while (System.currentTimeMillis() < startTime + timeout) {
@@ -812,6 +812,8 @@ public class AcmeFatUtils {
 		String serial2 = ((X509Certificate) endingCertificateChain[0]).getSerialNumber().toString(16);
 
 		assertThat("Expected a new certificate.", serial1, not(equalTo(serial2)));	
+		
+		return endingCertificateChain;
 	}
 	
 	/**

